src/core/functions/aggregate/llm_min_or_max.cpp

```diff
@@ -0,0 +1,210 @@
+
+    if (fixed_tokens > model_context_size) {
+        throw std::runtime_error("Fixed tokens exceed model context size");
+    }
+
+    available_tokens = model_context_size - fixed_tokens;
+}
+
+int LlmMinOrMax::calculateFixedTokens() const {
+    int fixed_tokens = 0;
+    fixed_tokens += Tiktoken::GetNumTokens(user_prompt);
+    fixed_tokens += Tiktoken::GetNumTokens(llm_min_or_max_template);
+    return fixed_tokens;
+}
+
+nlohmann::json LlmMinOrMax::GetElement(const nlohmann::json &tuples) {
+    inja::Environment env;
+    nlohmann::json data;
+    data["tuples"] = tuples;
+    data["user_prompt"] = user_prompt;
+    auto prompt = env.render(llm_min_or_max_template, data);
+
+    nlohmann::json settings;
+    auto response = ModelManager::CallComplete(prompt, model, settings);
+    return response["selected"];
+}
+
+nlohmann::json LlmMinOrMax::LlmMinOrMaxCall(nlohmann::json &tuples) {
+
+    if (tuples.size() == 1) {
+        return tuples[0]["content"];
+    }
+
+    auto num_tuples = tuples.size();
+    std::vector<int> batch_size_list;
+    auto used_tokens = 0;
+    auto batch_size = 0;
+    for (int i = 0; i < num_tuples; i++) {
+        used_tokens += Tiktoken::GetNumTokens(tuples[i].dump());
+        batch_size++;
+        if (used_tokens >= available_tokens) {
+            batch_size_list.push_back(batch_size);
+            used_tokens = 0;
+            batch_size = 0;
+        } else if (i == num_tuples - 1) {
+            batch_size_list.push_back(batch_size);
+        }
+    }
+
+    auto responses = nlohmann::json::array();
+    for (int i = 0; i < batch_size_list.size(); i++) {
+        auto start_index = i * batch_size;
+        auto end_index = start_index + batch_size;
+        auto batch = nlohmann::json::array();
+        for (int j = start_index; j < end_index; j++) {
+            batch.push_back(tuples[j]);
+        }
+        auto ranked_indices = GetElement(batch);
+        responses.push_back(batch[ranked_indices.get<int>()]);
+    }
+    return LlmMinOrMaxCall(responses);
+}
+
+// Static member initialization
+std::string LlmMinOrMaxOperation::model_name;
+std::string LlmMinOrMaxOperation::prompt_name;
+std::unordered_map<void *, std::shared_ptr<LlmMinOrMaxState>> LlmMinOrMaxOperation::state_map;
+
+
+void LlmMinOrMaxOperation::Initialize(const AggregateFunction &, data_ptr_t state_p) {
+    auto state_ptr = reinterpret_cast<LlmMinOrMaxState *>(state_p);
+
+    if (state_map.find(state_ptr) == state_map.end()) {
+        auto state = std::make_shared<LlmMinOrMaxState>();
+        state->Initialize();
+        state_map[state_ptr] = state;
+    }
+}
+
+void LlmMinOrMaxOperation::Operation(Vector inputs[], AggregateInputData &aggr_input_data, idx_t input_count, Vector &states,
+                      idx_t count) {
+    prompt_name = inputs[0].GetValue(0).ToString();
+    model_name = inputs[1].GetValue(0).ToString();
+
+    if (inputs[2].GetType().id() != LogicalTypeId::STRUCT) {
```

Comment:
I want to discuss 112 later. Separately, is there any value in having our checks for input as a function with a type, a Vector and error message?

Reply:
maybe if we unified our function arguments

Reply:
keep this laying around as a random idea and let's see if all input is currently already the same or not when it comes to these errors.